### ISSUE TITLE
FB14688465: ScrollView gesture ignored when `DragGesture` used by subview

### ISSUE THREAD

**Issue by Sjmarf** [feedback-assistant/reports#542]
- **Date:** 2024-08-06
- **Resolution:** Open
- **Area:** SwiftUI
- **OS:** iOS 18 dev beta 5
- **Type:** Incorrect/Unexpected Behavior
- **Keywords:** 

## Description

I have a view with a horizontal `DragGesture`. The `ScrollView` that my view is contained within ignores scroll up/down gestures. This behavior is new in iOS 18 (I have tested dev betas 4 & 5). Example code:

```swift
import SwiftUI

struct ContentView: View {
    var body: some View {
        ScrollView {
            ForEach(0..<20) { n in
                TileView(color: n % 2 == 0 ? .red : .blue)
            }
        }
    }
}

struct TileView: View {
    @GestureState var dragState: CGFloat = .zero
    
    let color: Color
    
    var body: some View {
        color
            .frame(height: 100)
            .gesture(
                DragGesture(
                    minimumDistance: 10,
                    coordinateSpace: .global
                )
                .updating($dragState) { value, state, _ in
                    state = value.translation.width
                }
            )
            .offset(x: dragState)
    }
}
```

See attached video of the above script running on iOS 17.5 (left) and iOS 18 dev beta 5 (right).

In the above example, the scroll gesture is always ignored. In my actual app, scroll gestures are only _sometimes_ ignored. 

Built using Xcode beta 5, happens on a sim (22A5326g) and on my physical device (22A5326f).

## Files

https://github.com/user-attachments/assets/a8923fce-2a92-4276-b351-83c327680a26

**Comment by emreertunc** [comment]
I have a similiar issue on my app. Found any solution to this?

**Comment by Sjmarf** [comment]
@emreertunc I did! 

I resorted to implementing a custom `UIGestureRecognizable` that wraps `UIPanGestureRecognizer`. Here's the code:

```swift
@available(iOS 18.0, *)
struct PanGesture: UIGestureRecognizerRepresentable {
    var handle: (UIPanGestureRecognizer) -> Void
    
    func makeCoordinator(converter: CoordinateSpaceConverter) -> Coordinator { .init() }
    
    func makeUIGestureRecognizer(context: Context) -> UIPanGestureRecognizer {
        let gesture = UIPanGestureRecognizer()
        gesture.delegate = context.coordinator
        gesture.isEnabled = true
        return gesture
    }
    
    func handleUIGestureRecognizerAction(_ recognizer: UIPanGestureRecognizer, context: Context) {
        handle(recognizer)
    }
    
    class Coordinator: NSObject, UIGestureRecognizerDelegate {
        func gestureRecognizer(
            _ gestureRecognizer: UIGestureRecognizer,
            shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer
        ) -> Bool {
            false
        }
        
        func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
            guard let panRecognizer = gestureRecognizer as? UIPanGestureRecognizer else { return false }

            let velocity = panRecognizer.velocity(in: gestureRecognizer.view)
            return abs(velocity.y) < abs(velocity.x)
        }
    }
}
```

```swift
.gesture(
    PanGesture { recognizer in
        // Do whatever you like with these values:
        let offset = recognizer.translation(in: recognizer.view).x
        let state = recognizer.state
    }
)
```

Because `UIGestureRecognizerRepresentable` is only available on iOS 18 and above, you'll need to fall back to the pure-SwiftUI solution on iOS 17 and earlier (if your app supports those versions). To do this, you can use something like this in the view body:

```swift
var body: some View {
    if #available(iOS 18.0, *) {
        bodyForVersion18
    } else {
        bodyForVersion17
    }
}

@available(iOS 18.0, *) @ViewBuilder
var bodyForVersion18: some View {
}

var bodyForVersion17: some View {
}
```

**Comment by emreertunc** [comment]
Yeah, it helped a lot. Actually, it took me a day to tweak and implement it into my project, but you saved the day! I hope more people stumble upon this issue and find the solution. If you'd like to see the app you helped with, you can check my page — there's a link to the Slean app.

Have a nice day!

**Comment by Sjmarf** [comment]
I'm glad you got it working! Good luck on your project :)

**Comment by utopyin** [comment]
> I hope this helps.

@Sjmarf hell yeah it helped man! Thank you so much.

I feel like this DragGesture+ScrollView behavior should be reported as a bug to Apple (especially when seeing the number of folks that had the same issue as us). This definitely doesn't look like a feature, more like a feature regression.

Anyway, you're solution is very elegant, and it actually made my implementation even cleaner.
I still have to support < iOS 18.0 though, which ends up adding quite a lot of complexity for a 'simple' custom swipable view :(

**Comment by Sjmarf** [comment]
@utopyin I'm glad it helped :)

I reported this to Apple via the feedback assistant during the iOS 18 beta, but never got a response.

**Comment by ajunjunandtmac** [comment]
> [@utopyin](https://github.com/utopyin) I'm glad it helped :)
> 
> I reported this to Apple via the feedback assistant during the iOS 18 beta, but never got a response.

Thanks for your workaround, it works, and also i wonder is there any progress made from Apple side?

**Comment by Sjmarf** [comment]
> Thanks for your workaround, it works, and also i wonder is there any progress made from Apple side?

You're welcome 🙂 And no, the feedback ticket is still open with 0 comments.

**Comment by ajunjunandtmac** [comment]
> > Thanks for your workaround, it works, and also i wonder is there any progress made from Apple side?
> 
> You're welcome 🙂 And no, the feedback ticket is still open with 0 comments.

That's so frustrated

**Comment by Sjmarf** [comment]
I have received a response from Apple:

> Thank you for filing this feedback report. We reviewed your report and determined the behavior you experienced is currently functioning as intended.

> When you add a DragGesture to a view inside a scroll view, it’s expected that this gesture takes priority over the scroll view’s own gestures. To add a DragGesture that doesn’t conflict with the containing scroll view’s own gestures, use a UILongPressGestureRecognizer within a UIGestureRecognizerRepresentable and return an appropriate value in its gestureRecognizer(_:shouldRecognizeSimultaneouslyWith:) delegate method, or configure it a way that it fails early if you can immediately determine that the gesture is invalid.

I assume they meant `UIPanGestureRecognizer` rather than than `UILongPressGestureRecognizer`. I've given an example implementation for this above.

**Comment by schen96** [comment]
Okay so for anyone still trying to figure this out, 

You need to set a larger minimumDistance, or otherwise the scrollView / draggesture competes against each other. Setting minimum distance to something like 30 and it works fine.
